src/control/server/ctl_storage_rpc.go

```diff
@@ -166,128 +170,330 @@ func (c *ControlService) scanScm(ctx context.Context, req *ctlpb.ScanScmReq) (*c
+			devicesToAdjust[rank] = &deviceSizeStat{
+				clusterPerTarget: math.MaxUint64,
+			}
+		}
+		devicesToAdjust[rank].devs = append(devicesToAdjust[rank].devs,
+			Device{
+				nvmeCtlr: ctlr,
+				smdIdx:   idx,
+			})
+		targetCount := uint64(len(dev.GetTgtIds()))
+		clusterPerTarget := dataClusterCount / targetCount
+		c.log.Debugf("SMD device %s (rank %d, ctlr %s) added to the list of device to adjust",
+			dev.GetUuid(), rank, ctlr.GetPciAddr())
+		if clusterPerTarget < devicesToAdjust[rank].clusterPerTarget {
+			c.log.Debugf("Updating number of clusters per target of rank %d: old=%d new=%d",
+				rank, devicesToAdjust[rank].clusterPerTarget, clusterPerTarget)
+			devicesToAdjust[rank].clusterPerTarget = clusterPerTarget
+		}
+	}
+
+	for _, ctlr := range resp.GetCtrlrs() {
+		engineCfg, err := c.getEngineCfgFromNvmeCtl(ctlr)
+		if err != nil {
+			c.log.Noticef("Skipping NVME controller %s: %s",
+				ctlr.GetPciAddr(), err.Error())
+			continue
+		}
+
+		for idx, dev := range ctlr.GetSmdDevices() {
+			rank := dev.GetRank()
+
+			if dev.GetDevState() != ctlpb.NvmeDevState_NORMAL {
+				c.log.Debugf("SMD device %s (rank %d, ctlr %s) not usable: device state %q",
+					dev.GetUuid(), rank, ctlr.GetPciAddr(), ctlpb.NvmeDevState_name[int32(dev.DevState)])
+				dev.TotalBytes = 0
+				dev.AvailBytes = 0
+				dev.UsableBytes = 0
+				continue
+			}
+
+			if dev.GetClusterSize() == 0 || len(dev.GetTgtIds()) == 0 {
+				c.log.Noticef("SMD device %s (rank %d,  ctlr %s) not usable: missing storage info",
+					dev.GetUuid(), rank, ctlr.GetPciAddr())
+				dev.AvailBytes = 0
+				dev.UsableBytes = 0
+				continue
+			}
+
+			if dev.GetRoleBits() != 0 && (dev.GetRoleBits()&storage.BdevRoleData) == 0 {
+				c.log.Debugf("SMD device %s (rank %d, ctlr %s) not used to store data (Role bits 0x%X)",
+					dev.GetUuid(), rank, ctlr.GetPciAddr(), dev.GetRoleBits())
+				dev.AvailBytes = 0
+				dev.UsableBytes = 0
+				continue
+			}
+
+			c.log.Debugf("Initial available size of SMD device %s (rank %d, ctlr %s): %s (%d bytes)",
+				dev.GetUuid(), rank, ctlr.GetPciAddr(), humanize.Bytes(dev.GetAvailBytes()), dev.GetAvailBytes())
+
+			clusterSize := uint64(dev.GetClusterSize())
+			availBytes := (dev.GetAvailBytes() / clusterSize) * clusterSize
+			if dev.GetAvailBytes() != availBytes {
+				c.log.Debugf("Adjusting available size of SMD device %s (rank %d, ctlr %s): from %s (%d Bytes) to %s (%d bytes)",
+					dev.GetUuid(), rank, ctlr.GetPciAddr(),
+					humanize.Bytes(dev.GetAvailBytes()), dev.GetAvailBytes(),
+					humanize.Bytes(availBytes), availBytes)
+				dev.AvailBytes = availBytes
+			}
+
+			dataClusterCount := dev.GetAvailBytes() / clusterSize
+			if dev.GetRoleBits() == 0 {
+				c.log.Debugf("No meta-data stored on SMD device %s (rank %d, ctlr %s)",
+					dev.GetUuid(), rank, ctlr.GetPciAddr())
+				addDeviceToAdjust(ctlr, idx, rank, dataClusterCount)
+				continue
+			}
+
+			getClusterCount := func(sizeBytes uint64, targetNb uint64) uint64 {
```

Comment:
Again, I think specifying helper functions at the file level rather than as function pointers inside another function would make this code easier to read.